NeosModLoader/Utility/EnumerableInjector.cs

```diff
@@ -0,0 +1,159 @@
+        /// </summary>
+        public Func<TIn, bool> PreItem
+        {
+            get => preItem;
+            set => preItem = value ?? throw new ArgumentNullException(nameof(value), "PreItem can't be null!");
+        }
+
+        /// <summary>
+        /// Gets called for each item to transform it, even if it won't be passed through.
+        /// </summary>
+        public Func<TIn, TOut> TransformItem
+        {
+            get => transformItem;
+            set => transformItem = value ?? throw new ArgumentNullException(nameof(value), "TransforItem can't be null!");
+        }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="EnumerableInjector{TIn, TOut}"/> class using the supplied input enumerable and transform function.
+        /// </summary>
+        /// <param name="enumerable">The enumerable to inject into and transform.</param>
+        /// <param name="transformItem">The transformation function.</param>
+        public EnumerableInjector(IEnumerable<TIn> enumerable, Func<TIn, TOut> transformItem)
+            : this(enumerable.GetEnumerator(), transformItem)
+        { }
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="EnumerableInjector{TIn, TOut}"/> class using the supplied input enumerator and transform function.
+        /// </summary>
+        /// <param name="enumerable">The enumerator to inject into and transform.</param>
+        /// <param name="transformItem">The transformation function.</param>
+        public EnumerableInjector(IEnumerator<TIn> enumerator, Func<TIn, TOut> transformItem)
+        {
+            this.enumerator = enumerator;
+            TransformItem = transformItem;
+        }
+
+        /// <summary>
+        /// Injects into and transforms the input enumeration.
+        /// </summary>
+        /// <returns>The injected and transformed enumeration.</returns>
+        public IEnumerator<TOut> GetEnumerator()
+        {
+            Prefix();
+
+            while (enumerator.MoveNext())
+            {
+                var item = enumerator.Current;
+                var returnItem = PreItem(item);
+                var transformedItem = TransformItem(item);
+
+                if (returnItem)
+                    yield return transformedItem;
+
+                PostItem(item, transformedItem, returnItem);
+            }
+
+            Postfix();
+        }
+
+        /// <summary>
+        /// Injects into and transforms the input enumeration without a generic type.
+        /// </summary>
+        /// <returns>The injected and transformed enumeration without a generic type.</returns>
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        private static void nothing()
+        { }
+
+        private static void nothing(TIn _, TOut __, bool ___)
+        { }
+
+        private static bool yes(TIn _) => true;
```

Comment:
I feel like these would be better to just be inline

Reply:
something something performance and less overhead

Reply:
...That sounds an awful lot like premature optimization, without any benchmarks to back it up, at the expense of readability ;'P

Reply:
I dunno, quite readable to me, and the end user won't care